packages/components/src/number-control/index.js

```diff
@@ -93,58 +93,47 @@ export function NumberControl(
 				nextValue = subtract( nextValue, incrementalValue );
 			}
 
-			nextValue = roundClamp( nextValue, min, max, incrementalValue );
-
-			state.value = nextValue;
+			state.value = constrainValue( nextValue );
 		}
 
 		/**
 		 * Handles drag to update events
 		 */
 		if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
-			const { delta, shiftKey } = payload;
-			const [ x, y ] = delta;
-			const modifier = shiftKey
+			const modifier = payload.shiftKey
 				? parseFloat( shiftStep ) * baseStep
 				: baseStep;
 
 			let directionModifier;
-			let directionBaseValue;
+			let delta;
 
 			switch ( dragDirection ) {
 				case 'n':
-					directionBaseValue = y;
+					delta = payload.delta[ 1 ];
 					directionModifier = -1;
 					break;
 
 				case 'e':
-					directionBaseValue = x;
+					delta = payload.delta[ 0 ];
 					directionModifier = isRTL() ? -1 : 1;
 					break;
 
 				case 's':
-					directionBaseValue = y;
+					delta = payload.delta[ 1 ];
 					directionModifier = 1;
 					break;
 
 				case 'w':
-					directionBaseValue = x;
+					delta = payload.delta[ 0 ];
 					directionModifier = isRTL() ? 1 : -1;
 					break;
 			}
 
-			const distance = directionBaseValue * modifier * directionModifier;
-			let nextValue;
-
-			if ( distance !== 0 ) {
-				nextValue = roundClamp(
-					add( currentValue, distance ),
-					min,
-					max,
-					modifier
-				);
+			if ( delta !== 0 ) {
+				delta = Math.ceil( Math.abs( delta ) ) * Math.sign( delta );
```

Comment:
The value of `delta` comes from react-use-gesture and is a decimal value. It's converted to an integer in order to preserve the `NumberControl`'s decimal value as entered (when `step="any"`). Done via `Math.ceil` etc. because single pixel movements produce values like: 
![image](https://user-images.githubusercontent.com/9000376/132381081-7552571a-96f7-411b-9c23-be2035262e8d.png)